### DISCUSSION TITLE
止盈止损单如何传递参数

### DISCUSSION THREAD

**Discussion by chwei2ch** [wondertrader/wtpy#55] [Q&A] [answered]
在下单时想直接设置止盈止损价格，如何传递参数？

**Comment by wondertrader** [comment, marked as answer]
WonderTrader的下单接口的逻辑，是参考MultiCharts来实现的，以CTA为例，一共提供了5个下单接口
```python
def stra_set_position(self, stdCode:str, qty:float, usertag:str = "", limitprice:float = 0.0, stopprice:float = 0.0):
    '''
    设置仓位
    @stdCode   合约/股票代码
    @qty    目标仓位，正为多仓，负为空仓
    @return 设置结果TRUE/FALSE
    '''
    self.__wrapper__.cta_set_position(self.__id__, stdCode, qty, usertag, limitprice, stopprice)
    

def stra_enter_long(self, stdCode:str, qty:float, usertag:str = "", limitprice:float = 0.0, stopprice:float = 0.0):
    '''
    多仓进场，如果有空仓，则平空再开多
    @stdCode   品种代码
    @qty    数量
    @limitprice 限价，默认为0
    @stopprice  止价，默认为0
    '''
    self.__wrapper__.cta_enter_long(self.__id__, stdCode, qty, usertag, limitprice, stopprice)

def stra_exit_long(self, stdCode:str, qty:float, usertag:str = "", limitprice:float = 0.0, stopprice:float = 0.0):
    '''
    多仓出场，如果剩余多仓不够，则全部平掉即可
    @stdCode   品种代码
    @qty    数量
    @limitprice 限价，默认为0
    @stopprice  止价，默认为0
    '''
    self.__wrapper__.cta_exit_long(self.__id__, stdCode, qty, usertag, limitprice, stopprice)

def stra_enter_short(self, stdCode:str, qty:float, usertag:str = "", limitprice:float = 0.0, stopprice:float = 0.0):
    '''
    空仓进场，如果有多仓，则平多再开空
    @stdCode   品种代码
    @qty    数量
    @limitprice 限价，默认为0
    @stopprice  止价，默认为0
    '''
    self.__wrapper__.cta_enter_short(self.__id__, stdCode, qty, usertag, limitprice, stopprice)

def stra_exit_short(self, stdCode:str, qty:float, usertag:str = "", limitprice:float = 0.0, stopprice:float = 0.0):
    '''
    空仓出场，如果剩余空仓不够，则全部平掉即可
    @stdCode   品种代码
    @qty    数量
    @limitprice 限价，默认为0
    @stopprice  止价，默认为0
    '''
    self.__wrapper__.cta_exit_short(self.__id__, stdCode, qty, usertag, limitprice, stopprice)
```
每个接口都有limitprice和stopprice两个参数
**参数的实现：**
当这两个参数不为0时，该信号作为条件单处理：即实时对比最新价格是否满足条件，如果满足再触发信号。因为CTAEngine是信号和执行剥离，所以这里的条件单都是本地通过实时tick触发的。
**参数的意义：**
limitprice，限价，这个比较好理解，一般来说，买入信号最终成交价格小于等于limitprice，卖出信号，最终成交价格大于等于limitprice；
stopprice，止价，也叫停损单，这个逻辑就和limitprice相反，买入信号会在最新价大于等于stopprice时成交，卖出信号会在最新价小于等于stopprice时成交。
为了方便理解，可以直观地认为，stop单就是价格突破的时候触发，limit单就是价格合适的时候再触发。
**参数用法的细节：**
* 如果下单接口同时传递了limitprice和stopprice，只会处理limitprice
* 当主K线闭合时，会清理掉所有的条件单。这样的好处是：*策略在oncalc中写触发逻辑的时候，不需要考虑太多状态相关的问题，设置条件单的代码不需要进行控制（这个逻辑和MC一致），因为底层已经处理好了*。
* 如果当前主K线周期内，有若干个待触发的条件单，当其中一个条件单触发以后，当前K线就不会再触发任何条件单了。这样的目的是：*防止重复发出信号*。

**Reply by chwei2ch** [reply]
感谢大佬，根据群里提示，举例如下：
假设为多单，现价5000，止损4500，止盈6000
stra_enter_long(stdCode, qty=10, usertag="enterlong", limitprice=0.0, stopprice=0.0)   开仓
stra_exit_long(stdCode, qty=10, usertag="exitlong", limitprice=0.0, stopprice=4500.0)  止损
stra_exit_long(stdCode, qty=10, usertag="exitlong", limitprice=6000.0, stopprice=0.0)  止盈

**Reply by wondertrader** [reply]
是的，就是这样的。
概括起来就是：stop单的方向应该和下单方向相同，limit方向应该和下单方向相反。

**Reply by qjaden** [reply]
这个条件单机制重启代码还有效吗

**Comment by kk19990709** [comment]
会不会加一些百分比的api会更好？比如涨10%止盈，跌20%止损
